lib/ddtrace/contrib/grpc/interception_context.rb

```diff
@@ -0,0 +1,47 @@
+require_relative 'datadog_interceptor'
+
+module GRPC
+  class InterceptionContext
+    # :nodoc:
+    # The `#intercept!` method is implemented in gRPC; this module
+    # will be prepended to the original class, effectively injecting
+    # our tracing middleware into the head of the call chain.
+    module InterceptWithDatadog
+      def intercept!(type, args = {})
+        if should_prepend?
+          datadog_interceptor = choose_datadog_interceptor(args)
+
+          @interceptors.unshift(datadog_interceptor.new) if datadog_interceptor
+
+          @trace_started = true
+        end
+
+        super
+      end
+
+      private
+
+      def should_prepend?
+        !trace_started? && !already_prepended?
+      end
+
+      def trace_started?
+        defined?(@trace_started) && @trace_started
+      end
+
+      def already_prepended?
+        @interceptors.any? do |interceptor|
+          interceptor.class.ancestors.include?(Datadog::Contrib::GRPC::DatadogInterceptor::Base)
+        end
+      end
+
+      def choose_datadog_interceptor(args)
+        if args.key?(:metadata)
```

Comment:
Can you explain the reasoning here for how an interceptor is chosen?

Reply:
In the gRPC Ruby core library, Client and Server interceptors have the same method names, but different keyword arguments. Clients always have a `metadata` keyword whereas Servers do not; servers always have a `call` keyword. A `metadata` is very similar to a header object in many HTTP libraries; it may be just a `Hash` in some cases. A `call` is an internal object ( a `GRPC::ActiveCall` ) that provides access to the ongoing communication between client and server; it provides access to the metadata sent by the client.

Reply:
I see, is there any more concrete/reliable way of differentiating between client and server calls? It'd be worthwhile if possible, otherwise we can just go with what will work.